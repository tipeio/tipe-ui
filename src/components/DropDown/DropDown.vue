<template>
  <div
    :data-tipe-ui="$options.name"
    :tabindex="tabindex"
  >
    <div @click="toggle">
      <slot
        v-bind="$props"
        name="control"
      />
    </div>
    <div
      v-if="open"
      class="panel-container"
      @click="toggle"
    >
      <slot
        v-bind="$props"
        name="panel"
      />
    </div>
  </div>
</template>

<script>
import inputProps from '@/types/InputProps'

export default {
  name: 'TipeDropDown',
  props: inputProps,
  data() {
    return {
      open: false
    }
  },
  methods: {
    toggle() {
      if (this.disabled || this.waiting) return
      this.open = !this.open
    }
  }
}
</script>

<style>
[data-tipe-ui='TipeDropDown'] {
  position: relative;
}

.panel-container {
  position: absolute;
  top: 100%;
  left: 0;
  background: transparent;
}
</style>
