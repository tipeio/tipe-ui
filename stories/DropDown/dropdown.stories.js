import { storiesOf } from '@storybook/vue'

import TipeDropDown from '../../src/components/DropDown'

const style = () => ({
  width: '200px'
})

storiesOf('DropDown', module)
  .add('default', () => ({
    components: { TipeDropDown },
    computed: { style },
    template: `
        <tipe-drop-down name="firstName">
          <div slot="control">hello world</div>
          <div slot="panel">goodbye world</div>
        </tipe-drop-down>`
  }))
  .add('disabled', () => ({
    components: { TipeDropDown },
    computed: { style },
    template: `
        <tipe-drop-down name="firstName" :disabled="false">
          <div slot="control" slot-scope="props">{{props.disabled}}</div>
          <div slot="panel">goodbye world</div>
        </tipe-drop-down>`
  }))
