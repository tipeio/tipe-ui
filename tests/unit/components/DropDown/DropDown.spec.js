import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import DropDown from '@/components/DropDown'

describe('DropDown', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const wrapper = shallowMount(DropDown)

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(DropDown)

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeDropDown')
    })

    it('<slot="control"> - renders control slot', () => {
      const slots = { control: '<div class="control"/>' }
      const wrapper = shallowMount(DropDown, { slots })

      expect(wrapper.contains('.control')).toEqual(true)
    })

    it('<slot="panel"> - panel slot not rendered', () => {
      const slots = { panel: '<div class="panel"/>' }
      const wrapper = shallowMount(DropDown, { slots })

      expect(wrapper.contains('.panel')).toEqual(false)
    })
  })

  describe('@event', () => {
    it('<slot="control" @click> - should toggle panel', () => {
      const slots = {
        control: '<div class="control"/>',
        panel: '<div class="panel"/>'
      }
      const wrapper = shallowMount(DropDown, { slots })

      wrapper.find('.control').trigger('click')
      expect(wrapper.contains('.panel')).toEqual(true)
      wrapper.find('.control').trigger('click')
      expect(wrapper.contains('.panel')).toEqual(false)
    })

    it('<slot="panel" @click> - should hide panel', () => {
      const slots = {
        control: '<div class="control"/>',
        panel: '<div class="panel"/>'
      }
      const wrapper = shallowMount(DropDown, { slots })

      wrapper.find('.control').trigger('click')
      expect(wrapper.contains('.panel')).toEqual(true)
      wrapper.find('.panel').trigger('click')
      expect(wrapper.contains('.panel')).toEqual(false)
    })
  })
})
