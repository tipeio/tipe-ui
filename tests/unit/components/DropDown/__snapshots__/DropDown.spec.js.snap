// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown <template> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeDropDown" tabindex="0">
  <div></div>
  <!---->
</div>
`;
